src/content/developers/tutorials/eip-1271-smart-contract-signatures/index.md

```diff
@@ -0,0 +1,123 @@
+
+> In fact, because EOA digital signatures are just based on cryptography, they can be generated and verified **off-chain**! This is how gasless DAO voting works — instead of submitting votes on-chain, digital signatures can be created and verified off-chain using cryptographic libraries.
+
+While your personal/EOA wallet has a private key, smart contracts do not have any sort of private or secret key (so "Log in with Ethereum", etc. cannot natively work with smart contracts).
+
+The problem EIP-1271 aims to solve: how can we tell that a smart contract signature is valid if the smart contract has no “secret” it can incorporate into the signature?
+
+## How does EIP-1271 work?
+
+Smart contracts don’t have private keys that can be used to sign messages. So how can we tell if a signature is authentic?
+
+Well, one idea is that we can just _ask_ the smart contract if a signature is authentic!
+
+What EIP-1271 does is it standardizes this idea “asking” a smart contract if a given signature is valid.
+
+A contract that implements EIP-1271 must have a function called `isValidSignature` which takes in a message and a signature. The contract can then run some validation logic (the spec does not enforce anything specific here) and then return a value indicating whether the signature is valid or not.
+
+If `isValidSignature` returns a valid result, that’s pretty much the contract saying “yes, I approve this signature + message!”
+
+### Interface
+
+Here’s the exact interface in the EIP-1271 spec (we’ll talk about the `_hash` parameter below, but for now, think of it as the message that is being verified):
+
+```jsx
+pragma solidity ^0.5.0;
+
+contract ERC1271 {
+
+  // bytes4(keccak256("isValidSignature(bytes32,bytes)")
+  bytes4 constant internal MAGICVALUE = 0x1626ba7e;
+
+  /**
+   * @dev Should return whether the signature provided is valid for the provided hash
+   * @param _hash      Hash of the data to be signed
+   * @param _signature Signature byte array associated with _hash
+   *
+   * MUST return the bytes4 magic value 0x1626ba7e when function passes.
+   * MUST NOT modify state (using STATICCALL for solc < 0.5, view modifier for solc > 0.5)
+   * MUST allow external calls
+   */
+  function isValidSignature(
+    bytes32 _hash,
+    bytes memory _signature)
+    public
+    view
+    returns (bytes4 magicValue);
+}
+```
+
+## Example EIP-1271 Implementation: Gnosis Safe
+
+Contracts can implement `isValidSignature` in many ways — the spec only doesn’t say much about the exact implementation.
+
+One notable contract which implements EIP-1271 is Gnosis Safe.
+
+In Gnosis Safe’s code, `isValidSignature` [is implemented](https://github.com/safe-global/safe-contracts/blob/main/contracts/handler/CompatibilityFallbackHandler.sol) so that signatures can be created and verified in [two ways](https://ethereum.stackexchange.com/questions/122635/signing-messages-as-a-gnosis-safe-eip1271-support):
```

Comment:
```suggestion
In Safe’s code, `isValidSignature` [is implemented](https://github.com/safe-global/safe-contracts/blob/main/contracts/handler/CompatibilityFallbackHandler.sol) so that signatures can be created and verified in [two ways](https://ethereum.stackexchange.com/questions/122635/signing-messages-as-a-gnosis-safe-eip1271-support):
```